libbeat/ml-importer/importer.go

```diff
@@ -121,3 +189,82 @@ func HaveXpackML(esClient MLLoader) (bool, error) {
 	}
 	return xpack.Features.ML.Available && xpack.Features.ML.Enabled, nil
 }
+
+// SetupModule creates ML jobs, data feeds and dashboards for modules.
+func SetupModule(kibanaClient MLSetupper, module, prefix string) error {
+	setupURL := fmt.Sprintf(kibanaSetupModuleURL, module)
+	prefixPayload := fmt.Sprintf("{\"prefix\": \"%s\"}", prefix)
+	status, response, err := kibanaClient.Request("POST", setupURL, nil, strings.NewReader(prefixPayload))
+	if status != 200 {
+		return errors.Errorf("cannot set up ML with prefix: %s", prefix)
+	}
+	if err != nil {
+		return err
+	}
+
+	return checkResponse(response)
+}
+
+func checkResponse(r []byte) error {
+	var errs multierror.Errors
+
+	var resp MLResponse
+	err := json.Unmarshal(r, &resp)
+	if err != nil {
+		return err
+	}
+
+	for _, feed := range resp.Datafeeds {
+		if !feed.Success {
+			if strings.HasPrefix(feed.Error.Msg, "[resource_already_exists_exception]") {
```

Comment:
My intuition here is that these types of checks are prone to breakages. I don't know how well defined the API spec is for this interface, but my guess is that error message is not documented to begin with this string so that could change.

I think it's OK to do these types of checks when there are no other options, but we should have tests in place to detect upstream changes in the error messages. Are there already tests for these cases? I'm not seeing any coverage in the report: https://codecov.io/gh/elastic/beats/pull/6604/diff#D1-219

Reply:
I agree, this part is quite dirty. Unfortunately, I haven't been able to find anything else which would tell the reason of the failure.

I will add tests to avoid breaking anything.